ir/src/main/scala/org/scalajs/core/ir/Printers.scala

```diff
@@ -1052,4 +1052,601 @@ object Printers {
+        indent()
+        var rest = ts
+        while (rest.nonEmpty) {
+          print(rest.head)
+          rest = rest.tail
+          if (rest.nonEmpty)
+            println()
+        }
+        undent()
+        paren()
+      }
+    }
+
+    private final def printlnList(ts: List[Tree]): Unit = {
+      printList(ts)
+      println()
+    }
+
+    private final def printIdentList(is: List[Ident]): Unit = {
+      if (is.isEmpty) {
+        print("List ()")
+      } else {
+        println("List (")
+        indent()
+        var rest = is
+        while (rest.nonEmpty) {
+          print(rest.head)
+          rest = rest.tail
+          if (rest.nonEmpty)
+            println()
+        }
+        undent()
+        paren()
+      }
+    }
+
+    private final def printlnIdentList(is: List[Ident]): Unit = {
+      printIdentList(is)
+      println()
+    }
+
+    private final def print(ident: Ident): Unit =
+      print(s"Ident(${ident.name}, ${ident.originalName})")
+
+    private final def println(ident: Ident): Unit = {
+      print(ident)
+      out.write("\n")
+    }
+
+    private def print(tpe: Type): Unit =
+      print(tpe.toString)
+
+
+    private final def println(tp: Type): Unit = {
+      print(tp)
+      out.write("\n")
+    }
+
+    private final def paren(): Unit = {
+      out.write(")")
+    }
+
+    private final def print(tree: Tree): Unit = {
+      tree match {
+        case EmptyTree =>
+          print("EmptyTree")
+
+        // Definitions
+
+        case VarDef(ident, vtpe, mutable, rhs) =>
+          println("VarDef (")
+          indent()
+          println(ident)
+          println(vtpe)
+          println("mutable = " + mutable.toString)
+          print(rhs); paren()
+          undent()
+
+
+        case ParamDef(ident, ptpe, mutable, rest) =>
+          println("ParamDef (")
+          indent()
+          println(ident)
+          println(ptpe)
+          println("mutable = " + mutable.toString)
+          print("rest = " + rest.toString); paren()
+          undent()
+
+        // Control flow constructs
+
+        case Skip() =>
+          print("Skip")
+
+        case Block(trees) =>
+          printList(trees)
+
+        case Labeled(label, tpe, body) =>
+          println("Labeled (")
+          indent()
+          println(tpe)
+          print(body); paren()
+          undent()
+
+        case Assign(lhs, rhs) =>
+          println("Assign (")
+          indent()
+          println(lhs)
+          print(rhs); paren()
+          undent()
+
+        case Return(expr, label) =>
+          println("Return (")
+          indent()
+          println(expr)
+          label match {
+            case Some(l) => print(l)
+            case None    => print("None")
+          }
+          paren()
+          undent()
+
+        case If(cond, thenp, elsep) =>
+          println("If (")
+          indent()
+          println(cond)
+          println(thenp)
+          print(elsep); paren()
+          undent()
```

Comment:
For trees like `If` where the `tpe` is an explicit parameter to the constructor of `If` (as opposed to being inferred), I would suggest you also print it explicitly, so that an `If` looks like this:
```scala
If(
  BooleanLiteral(true),
  IntLiteral(2),
  IntLiteral(3))(
  IntType)
```